dfusion_rust_core/src/models/solution.rs

```diff
@@ -17,12 +17,81 @@ pub struct Solution {
 impl Solution {
     pub fn trivial(num_orders: usize) -> Self {
         Solution {
-            objective_value: Some(U256::zero()),
             prices: vec![0; TOKENS as usize],
             executed_buy_amounts: vec![0; num_orders],
             executed_sell_amounts: vec![0; num_orders],
         }
     }
+
+    /// Returns the token price for token or `None` if the token is invalid.
+    fn price(&self, token: u16) -> Option<u128> {
+        self.prices.get(token as usize).cloned()
+    }
+
+    /// Compute the result of the objective function for the solution. This is
+    /// currently defined as *U - dUT* where *U* is the total utility of the
+    /// trade and *dUT* is the total disreguarded utility of touched orders.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` for any solution that is positive valid solution.
+    /// Returns `None` for invalid solutions or orders. A solution might be
+    /// invalid if:
+    /// - The number of orders do not match the solution buy/sell amounts
+    /// - The solution values lead to invalid utility or disregarded utility
+    ///   results
+    /// - The orders reference tokens that are not priced by the solution
+    ///
+    /// Note that just because this method returns some value does not mean that
+    /// the solution is actually valid WRT to the smart contract as it currently
+    /// does not verify that things like `execSellAmount * p[sellToken] * (1-phi)
+    /// == execBuyAmount * p[buyToken]` or token conservation hold.
+    pub fn objective_value(&self, orders: &[Order]) -> Option<U256> {
+        if orders.len() != self.executed_buy_amounts.len()
+            || orders.len() != self.executed_sell_amounts.len()
+        {
+            return None;
+        }
+
+        // note that we don't actually have to do the same checking here (such as
+        // account balances and such) as we do in the smart contract - we only
+        // care if the solution is valid given the provided orders
+
+        let mut u = U256::zero();
+        let mut du = U256::zero();
+
+        let mut fee_buy_amount = U256::zero();
+        let mut fee_sell_amount = U256::zero();
+
+        for (i, order) in orders.iter().enumerate() {
+            let buy_price = self.price(order.buy_token)?;
+            let sell_price = self.price(order.sell_token)?;
+            let exec_buy_amount = self.executed_buy_amounts[i];
+            let exec_sell_amount = self.executed_sell_amounts[i];
+
+            if exec_buy_amount == 0 && exec_sell_amount == 0 {
+                // this order was not touched, so skip
+                continue;
+            }
+
+            u = u.checked_add(order.utility(buy_price, exec_buy_amount, exec_sell_amount)?)?;
+            du = du.checked_add(order.disregarded_utility(
+                buy_price,
+                sell_price,
+                exec_sell_amount,
+            )?)?;
+
+            if order.buy_token == 0 {
```

Comment:
Is this correct? We would only count a fee if we are trading with the fee token (buying or selling it). While this is a restriction that the naive solver does, the "real" solver would also trade non-fee token with each other while subtracting a fee. I think we might have to compute the token conservation excess for the fee token and half that value.

Reply:
I'm not sure... I was hoping it was. Maybe I tried to do what `TokenConservation.sol` but only when it affects the token conservation for the fee token. I might have missed something though.

Reply:
Right now I don't think we accrue any fee in this logic if we are trading e.g. token1 and token2. However, in reality that also incurs a few (which just gets pushed around and eventually converted in token0)

Reply:
I am a little confused about how the fee token conservation works. From the contract, token conservation is calculated here:
```
tokenConservation.updateTokenConservation(
    order.buyToken,
    order.sellToken,
    tokenIdsForPrice,
    executedBuyAmount,
    executedSellAmount
);
```
and the calculation looks like this:
```
function updateTokenConservation(
    int[] memory self,
    uint16 buyToken,
    uint16 sellToken,
    uint16[] memory tokenIdsForPrice,
    uint128 buyAmount,
    uint128 sellAmount
) internal pure {
    uint buyTokenIndex = findPriceIndex(buyToken, tokenIdsForPrice);
    uint sellTokenIndex = findPriceIndex(sellToken, tokenIdsForPrice);
    self[buyTokenIndex] = self[buyTokenIndex].sub(int(buyAmount));
    self[sellTokenIndex] = self[sellTokenIndex].add(int(sellAmount));
} 
```

and from what I understood `{buy,sell}TokenIndex = 0` and `{buy,sell}Token = 0` for the fee token, that is why I wrote it the way I did.

I don't really understand how the calculation works.

Reply:
The other thing I can work from is: `exec_sell_amount * sell_price * (1 - phi) = exec_buy_amount * buy_price` so would the fee just be: `exec_buy_amount * buy_price - exec_sell_amount * sell_price` converted into token0?